astropy/io/fits/hdu/image.py

```diff
@@ -245,7 +282,26 @@ def data(self, data):
         else:
             raise ValueError('not a valid data array')
 
+        # Update the header, including adding BZERO/BSCALE if new data is
+        # unsigned. Does not change the values of self._bitpix,
+        # self._orig_bitpix, etc.
         self.update_header()
+        if (data is not None and was_unsigned):
+            self._update_header_scale_info(data.dtype)
```

Comment:
Right; this ends up deleting those keywords only when:

+ The data before this set was unsigned, and 
+ The new data is *not* unsigned.

So BZERO and BSCALE are present when the data is unsigned but are removed when the data is set to something whose type is *not* unsigned.